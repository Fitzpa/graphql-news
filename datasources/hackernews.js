import { RESTDataSource } from 'apollo-datasource-rest';

export class HackerNewsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = 'https://hacker-news.firebaseio.com/v0/';
  }

  articleReducer({ id, by, url, time, title } = {}) {
    return {
      id: `hn-${id}`,
      title,
      author: by,
      url,
      time,
      source: 'HackerNews',
    };
  }

  async getAllArticleIds() {
    const result = await this.get('topstories.json');
    return result;
  }

  async getArticle(articleId) {
    const result = await this.get(`item/${articleId}.json`);
    return this.articleReducer(result);
  }

  getArticlesByIds(articleIds) {
    return Promise.all(articleIds.map(articleId => this.getArticle(articleId)));
  }

  async getAllArticles() {
    let articleIds = await this.getAllArticleIds();
    articleIds = articleIds.slice(0, 100);
    return Promise.all(articleIds.map(articleId => this.getArticle(articleId)));
  }
}