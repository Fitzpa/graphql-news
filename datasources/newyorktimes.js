import { RESTDataSource } from 'apollo-datasource-rest';

export class NewYorkTimesAPI extends RESTDataSource {
  articleReducer({ id, byline, url, published_date, title } = {}) {
    return {
      id: `nyt-${id}`,
      title,
      author: byline,
      url,
      time: published_date,
      source: 'New York Times',
    };
  }

  async getAllArticles() {
    const result = await this.get(
      'https://api.nytimes.com/svc/mostpopular/v2/viewed/1.json?api-key=Jg1WQqSCu08PZKrQyqPGNN7F7FCYVnxp'
    );
    return result?.results?.map(article => this.articleReducer(article));
  }
}
