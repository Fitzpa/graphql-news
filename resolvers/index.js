export const resolvers = {
  Query: {
    allArticles: (_, __, { dataSources }) =>
      Promise.all(
        Object.keys(dataSources).map(source =>
          dataSources[source].getAllArticles()
        )
      ).then(result => result.reduce((acc, data) => acc.concat(data), [])),

    allArticlesBySource: (_, { source }, { dataSources }) =>
      dataSources[source].getAllArticles(),

    articleBySource: (_, { id, source }, { dataSources }) =>
      dataSources[source].getArticle(id, source),

    articlesBySource: (_, { ids, source }, { dataSources }) =>
      dataSources[source].getArticlesByIds(ids),
  },
};
