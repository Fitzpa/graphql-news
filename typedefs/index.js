import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    allArticles: [Article!]!
    allArticlesBySource(source: String!): [Article!]!
    articleBySource(id: ID!, source: String!): Article
    articlesBySource(ids: [Int!]!, source: String!): [Article!]!
  }

  type Article {
    id: ID!
    title: String
    author: String
    url: String
    time: String
    source: String
  }
`;
