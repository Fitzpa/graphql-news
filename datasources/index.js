import { HackerNewsAPI } from './hackernews';
import { NewYorkTimesAPI } from './newyorktimes';

export default {
  hackernews: new HackerNewsAPI(),
  newyorktimes: new NewYorkTimesAPI(),
};
