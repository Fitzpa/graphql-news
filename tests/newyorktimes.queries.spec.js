import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';
import { NewYorkTimesAPI } from '../datasources/newyorktimes';
import { resolvers } from '../resolvers';
import { typeDefs } from '../typedefs';
import { getArticlePreReducerStub } from '../fixtures/newyorktimes';

const GET_ALL_ARTICLES_BY_SOURCE = gql`
  query getAllArticlesBySource($source: String!) {
    allArticlesBySource(source: $source) {
      id
      title
      author
      url
      time
      source
    }
  }
`;

const constructTestServer = () => {
  const newyorktimesAPI = new NewYorkTimesAPI();

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      newyorktimes: newyorktimesAPI,
    }),
  });

  return { server, newyorktimesAPI };
};

describe('[Queries.NewYorkTimesAPI]', () => {
  it('fetches an array of articles from the New York Times API', async () => {
    const { server, newyorktimesAPI } = constructTestServer();

    newyorktimesAPI.get = jest.fn(() => ({
      results: [getArticlePreReducerStub],
    }));

    const { query } = createTestClient(server);

    const res = await query({
      query: GET_ALL_ARTICLES_BY_SOURCE,
      variables: { source: 'newyorktimes' },
    });

    expect(res).toMatchSnapshot();
  });
});
