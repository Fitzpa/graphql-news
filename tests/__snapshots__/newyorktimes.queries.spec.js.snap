// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Queries.NewYorkTimesAPI] fetches an array of articles from the New York Times API 1`] = `
Object {
  "data": Object {
    "allArticlesBySource": Array [
      Object {
        "author": "By NICHOLAS FANDOS",
        "id": "nyt-100000006666678",
        "source": "New York Times",
        "time": "2019-10-02",
        "title": "Welcome to Estoniaâ€™s Isle of Women",
        "url": "https://www.nytimes.com/2019/10/02/travel/kihnu.html",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
